### PR TITLE
nuttx/drivers: add regmap subsystems support.

### PR DESCRIPTION
**Pull request by Otpvondoiats**
regmap: add i2c and spi bus support.

## Summary
At this stage, it supports 8bit use, and multi-bytes need to be converted by yourself. Multi-bit formatting will be added later。
## Impact
None
## Testing
Ubuntu 20.04 
BUS:I2c/SIM 
sensor: BMI160

### CODE REVIEW

**Comment by acassis** [comment]
@Otpvondoiats please include some basic Documentation/ explaining how to use it.

File: drivers/regmap/internal.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+typedef CODE void (*regmap_lock_t)(FAR void *);
+typedef CODE void (*regmap_unlock_t)(FAR void *);
+
+/* Configuration for the register map of a device.
+ * This structure is only used inside regmap.
+ */
+
+struct regmap_s
+{
+  mutex_t mutex;
```
Comment (anchao) [inline drivers/regmap/internal.h:48]:
```suggestion
  mutex_t *mutex;
```
alloc the resource only if mutex lock is needed

**Comment by Otpvondoiats** [comment]
> @Otpvondoiats please include some basic Documentation/ explaining how to use it.

Done

File: drivers/regmap/regmap.c [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+ *
+ ****************************************************************************/
+
+int regmap_bulk_write(FAR struct regmap_s *map, unsigned int reg,
+                      FAR const void *val, unsigned int val_count)
+{
+  size_t val_bytes = map->val_bytes;
+  unsigned int ival;
+  FAR uint8_t *ptr;
+  int ret = 0;
```
Comment (anchao) [inline drivers/regmap/regmap.c:203]:
```suggestion
  int ret = -ENOSYS;
```

File: drivers/regmap/regmap.c [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+            ival = *(FAR uint8_t *)ptr;
+            break;
+          case 2:
+            ival = *(FAR uint16_t *)ptr;
+            break;
+          case 4:
+            ival = *(FAR uint32_t *)ptr;
+            break;
+          default:
+            goto out;
```
Comment (anchao) [inline drivers/regmap/regmap.c:230]:
```suggestion
            ret = -EINVAL;
            goto out;
```

File: drivers/regmap/regmap.c [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+ ****************************************************************************/
+
+int regmap_bulk_read(FAR struct regmap_s *map, unsigned int reg,
+                     FAR void *val, unsigned int val_count)
+{
+  FAR uint32_t *u32 = val;
+  FAR uint16_t *u16 = val;
+  FAR uint8_t  *u8  = val;
+  unsigned int ival;
+  int ret = 0;
```
Comment (anchao) [inline drivers/regmap/regmap.c:309]:
```suggestion
  int ret = -ENOSYS;
```

File: drivers/regmap/regmap.c [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+  int i;
+
+  DEBUGASSERT(REGMAP_ALIGNED(reg, map->reg_stride));
+
+  if (map->read != NULL)
+    {
+      map->lock(map);
```
Comment (anchao) [inline drivers/regmap/regmap.c:316]: move map->lock/unlock out of if/else block

File: drivers/regmap/regmap.c (hunk 2 above)
Comment (Otpvondoiats) [reply drivers/regmap/regmap.c]: Done.

File: drivers/regmap/regmap.c (hunk 3 above)
Comment (Otpvondoiats) [reply drivers/regmap/regmap.c:232]: Done.

File: drivers/regmap/regmap.c (hunk 4 above)
Comment (Otpvondoiats) [reply drivers/regmap/regmap.c]: Done.

File: drivers/regmap/regmap.c (hunk 5 above)
Comment (Otpvondoiats) [reply drivers/regmap/regmap.c]: Done.

**Review by anchao — APPROVED** [review]

File: drivers/regmap/regmap.c [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+    }
+
+  if (config->disable_locking)
+    {
+      map->lock   = regmap_lock_unlock_none;
+      map->unlock = regmap_lock_unlock_none;
+    }
+  else
+    {
+      map->mutex = kmm_zalloc(sizeof(*map->mutex));
```
Comment (anchao) [inline drivers/regmap/regmap.c:112]:
```suggestion
      map->mutex = kmm_zalloc(sizeof(*map->mutex));
      if (map->mutex == NULL)
        {
          kmm_free(map);
          return NULL;
        }
```

File: drivers/regmap/regmap.c [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+ *
+ * Assumptions/Limitations:
+ *   None.
+ ****************************************************************************/
+
+void regmap_exit(FAR struct regmap_s *map)
+{
+  if (!map->disable_locking)
+    {
+      nxmutex_destroy(map->mutex);
```
Comment (anchao) [inline drivers/regmap/regmap.c:372]:
```suggestion
      nxmutex_destroy(map->mutex);
      kmm_free(map->mutex);
```

File: drivers/regmap/regmap.c (hunk 6 above)
Comment (Otpvondoiats) [reply drivers/regmap/regmap.c:112]: Done

**Review by anchao — APPROVED** [review]

File: drivers/regmap/internal.h [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+typedef CODE void (*regmap_lock_t)(FAR void *);
+typedef CODE void (*regmap_unlock_t)(FAR void *);
+
+/* Configuration for the register map of a device.
+ * This structure is only used inside regmap.
+ */
+
+struct regmap_s
+{
+  FAR mutex_t *mutex;
```
Comment (xiaoxiang781216) [inline drivers/regmap/internal.h:48]: let's move to the end of struct to avoid malloc twice

Comment (Otpvondoiats) [reply drivers/regmap/internal.h]: Done

**Comment by Otpvondoiats** [comment]
s

File: drivers/regmap/internal.h [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+  /* The register address stride. Valid register addresses are a multiple
+   * of this value. If set to 0, a value of 1 will be used.
+   */
+
+  int reg_stride;
+
+  /* Prevent fragmentation */
+
+  FAR mutex_t *mutex;
```
[the remark this answers is not in the document]
Comment (Otpvondoiats) [reply drivers/regmap/internal.h]: Done

**Review by xiaoxiang781216 — APPROVED** [review]

File: drivers/regmap/regmap.c [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+ ****************************************************************************/
+
+static void regmap_lock_unlock_none(FAR void *context)
+{
+}
+
+static void regmap_lock_mutex(FAR void *context)
+{
+  FAR struct regmap_s *map = context;
+  nxmutex_lock(&map->mutex[0]);
```
Comment (xiaoxiang781216) [inline drivers/regmap/regmap.c:53]:
```suggestion
  nxmutex_lock(map->mutex);
```

**Pull request merged**